OpenRA.Mods.Common/Traits/AutoTarget.cs

```diff
@@ -351,74 +395,63 @@ Target ChooseTarget(Actor self, AttackBase ab, PlayerRelationship attackStances,
+				{
+					if (attackStances == PlayerRelationship.Enemy && self.Owner.RelationshipWith(frozen.Owner) == PlayerRelationship.Ally)
+						continue;
+
+					// Bot-controlled units aren't yet capable of understanding visibility changes
+					// Prevent that bot-controlled units endlessly fire at frozen actors.
+					// TODO: Teach the AI to support long range artillery units with units that provide line of sight
+					if (self.Owner.IsBot && frozen.Actor == null)
+						continue;
+
+					targetsInRange.Add(Target.FromFrozenActor(frozen));
+				}
+			}
 
 			foreach (var target in targetsInRange)
 			{
 				BitSet<TargetableType> targetTypes;
 				Player owner;
 				if (target.Type == TargetType.Actor)
 				{
-					// PERF: Most units can only attack enemy units. If this is the case but the target is not an enemy, we
-					// can bail early and avoid the more expensive targeting checks and armament selection. For groups of
-					// allied units, this helps significantly reduce the cost of auto target scans. This is important as
-					// these groups will continuously rescan their allies until an enemy finally comes into range.
-					if (attackStances == PlayerRelationship.Enemy && !target.Actor.AppearsHostileTo(self))
-						continue;
-
-					// Check whether we can auto-target this actor
 					targetTypes = target.Actor.GetEnabledTargetTypes();
-
-					if (PreventsAutoTarget(self, target.Actor) || !target.Actor.CanBeViewedByPlayer(self.Owner))
-						continue;
-
 					owner = target.Actor.Owner;
 				}
 				else if (target.Type == TargetType.FrozenActor)
 				{
-					if (attackStances == PlayerRelationship.Enemy && self.Owner.RelationshipWith(target.FrozenActor.Owner) == PlayerRelationship.Ally)
-						continue;
-
-					// Bot-controlled units aren't yet capable of understanding visibility changes
-					// Prevent that bot-controlled units endlessly fire at frozen actors.
-					// TODO: Teach the AI to support long range artillery units with units that provide line of sight
-					if (self.Owner.IsBot && target.FrozenActor.Actor == null)
-						continue;
-
 					targetTypes = target.FrozenActor.TargetTypes;
 					owner = target.FrozenActor.Owner;
 				}
 				else
 					continue;
 
-				var validPriorities = activePriorities.Where(ati =>
-				{
-					// Already have a higher priority target
-					if (ati.Priority < chosenTargetPriority)
-						return false;
-
-					// Incompatible relationship
-					if (!ati.ValidRelationships.HasRelationship(self.Owner.RelationshipWith(owner)))
-						return false;
-
-					// Incompatible target types
-					if (!ati.ValidTargets.Overlaps(targetTypes) || ati.InvalidTargets.Overlaps(targetTypes))
-						return false;
-
-					return true;
-				}).ToList();
-
-				if (validPriorities.Count == 0)
+				var targetPriotity = GetMaxTargetPriority(self.Owner, owner, targetTypes, chosenTargetPriority);
+				if (targetPriotity == int.MinValue)
 					continue;
 
 				// Make sure that we can actually fire on the actor
```

Comment:
the below code could also be optimized.

```csharp
// Make sure that we can actually fire on the actor
if (allowMove)
{
	if (ab.ChooseArmamentsForTarget(target, false).Any())
		continue;
}
else
{
	if (ab.ChooseArmamentsForTarget(target, false).Any(arm =>
		target.IsInRange(self.CenterPosition, arm.MaxRange()) &&
		!target.IsInRange(self.CenterPosition, arm.Weapon.MinRange)))
		continue;
}
```

The first part doesn't allocate memory, the second part doesn't need to parse all targets